### ISSUE TITLE
Colors get washed out on HDR display (Windows)

### ISSUE THREAD

**Issue by DonutWrecker** [godotengine/godot#79700]
### Godot version

4.2.dev1

### System information

Godot v4.2.dev1 - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1650 (NVIDIA; 31.0.15.3667) - AMD Ryzen 3 3100 4-Core Processor (8 Threads)

### Issue description

I'm having the issue on my main display, a 4K/60Hz/HDR10 monitor. I know it's HDR is working properly since I'm playing games on fullscreen with HDR on all the time.

The issue has a strange behavior. When running the project on editor (2D) on exclusive fullscreen, colors get washed out (HDR gets disabled?), after a few seconds of gameplay. The colors look exactly like on regular fullscreen and windowed (HDR off). 

The issue is also present on Compatibility mode. I remember the same issue was present on v4.0.3-stable as well.

Having tried all other combinations, only exclusive fullscreen presents a stutter-free gameplay, (vsync on, fps unlocked/or locked to 60), that's why I'm using exclusive fullscreen instead of regular fullscreen. Regular fullscreen with FPS limit 60 causes stutter regardless of vsync.

### Steps to reproduce

1. Run the test project.
2. Watch tiles closely, the ground specifically.
3. After a few seconds (5-30 seconds, mostly), the screen will blink for a moment, and you'll see the tiles' colors getting washed out.

### Minimal reproduction project

[test_project.zip](https://github.com/godotengine/godot/files/12105013/test_project.zip)

**Comment by clayjohn** [comment]
Godot doesn't currently support HDR screens right now. It always renders in LDR. Likely what you are seeing is the result of sending the display an image that is intended for LDR (8 bit color, normalized range) when the display is expecting something with a higher range and bit depth

**Comment by DonutWrecker** [comment]
> Godot doesn't currently support HDR screens right now. It always renders in LDR. Likely what you are seeing is the result of sending the display an image that is intended for LDR (8 bit color, normalized range) when the display is expecting something with a higher range and bit depth

Oh, I see. Well, hopefully it'll get HDR support soon.

**Comment by DonutWrecker** [comment]
> This is a Windows issue we can't fix when SDR content is rendered on a display with HDR enabled.
> 
> You can somewhat improve the situation on Windows 11 using the official HDR calibration app (available on the Microsoft store), but it's not available on Windows 10 to my knowledge.
> 
> > I know it's HDR is working properly since I'm playing games on fullscreen with HDR on all the time.
> 
> Until Windows gets its SDR tonemapping right, [it's strongly recommended not to keep HDR enabled all the time](https://tftcentral.co.uk/articles/heres-why-you-should-only-enable-hdr-mode-on-your-pc-when-you-are-viewing-hdr-content). Windows 11 has understood this by adding a Windows + Alt + B keyboard shortcut to toggle HDR. On Windows 10, see the script in [this forum thread](https://www.tenforums.com/graphic-cards/155095-windows-10-hdr-toggle-shortcut.html). The general recommendation for HDR usage is to use Windows 11 still, as it comes with a _lot_ of improvements regarding HDR support (even though it still leaves to be desired).
> 
> PS: What monitor model do you have? Some displays fare better than others when it comes to displaying SDR content in HDR, but it also depends on your configuration ([DTM vs HGIG in the monitor settings, with HGIG being closer to artistic intent](https://www.reddit.com/r/OLED_Gaming/comments/ll63oj/hgig_or_dynamic_tone_mapping/)).

In my case, HDR is only triggered when a fullscreen app launches, in my case, only games, and pretty much every game I've ever played.

It's interesting, because I've never felt an urge to turn-off HDR on any of the games I've played, both 2D and 3D, or old/new, ever. Everything looked alright. If I'm going to publish a game, I expect it to run as smooth and out-of-the-box as any of the games I've played. Am I missing something?

**Comment by Calinou** [comment]
> If I'm going to publish a game, I expect it to run as smooth and out-of-the-box as any of the games I've played. Am I missing something?

Unfortunately, it's not as clear and cut when it comes to Windows HDR. Only macOS has really gotten HDR truly right from an implementation perspective. (Support for HDR on Linux is barely a thing right now.)

**Comment by DonutWrecker** [comment]
> > If I'm going to publish a game, I expect it to run as smooth and out-of-the-box as any of the games I've played. Am I missing something?
> 
> Unfortunately, it's not as clear and cut when it comes to Windows HDR. Only macOS has really gotten HDR truly right from an implementation perspective. (Support for HDR on Linux is barely a thing right now.)
> 
> If you're wondering why HDR adoption isn't that high as of 2023, it's a big reason why (on top of the hardware often not being up to par) 🙂
> 
> > It's interesting, because I've never felt an urge to turn-off HDR on any of the games I've played, both 2D and 3D, or old/new, ever. Everything looked alright.
> 
> Have you tried other games running OpenGL or Vulkan? Direct3D apps likely get different treatment.
> 
> [Special K](https://special-k.info/) may also be worth a look into:
> 
> > HDR retrofit for OpenGL, Vulkan, Direct3D 11 and 12 titles
> 
> I'd be interested to know whether it works on Godot, but I'm not on my Windows setup right now.
> 
> Note that #70315 could make it possible to benefit from Auto HDR on Windows 11 (if enabled in the Windows settings), although I haven't gotten it to work on my end. This is likely because Auto HDR needs to be enabled on a per-game basis by Microsoft, and I'm not sure if you can override this.

I can't say for sure, just yet. I have to test on several different games. I'll update the issue as soon as possible. I've just looked up some of the indie games I've played recently, all of those have an DirectX requirement. You may want to consider implementing Direct3D for Godot, since Unix-based systems, except consoles, will never be the main gaming platform.

**Comment by YuriSizov** [comment]
It is working perfectly fine in apps that support HDR. But Windows environment itself does not.

**Comment by DonutWrecker** [comment]
I want to clarify some things before I get into my test results. By saying "washed-out", I don't mean the colors get messed up completely, or they look super disturbing. I mean the colors get less saturated, or less vibrant. It is a subtle change, but recognizable.

The other is, I may be wrong about the HDR is being triggered when a fullscreen app launches. Upon further inspection, I've seen that the HDR pop-up is presented when I escape from the fullscreen. The screen gets black for a second when I enable fullscreen, but with no pop-up; and when I escape the fullscreen, the screen gets black for a second again, then the HDR pop-up is seen. I may not have given much attention to it, I'm sorry about that.

I got to test it out with some other made-with-Godot games. 2 games exactly, and they're both made with Godot 3 or 2 probably.

First game I've tried was [Bottomless](https://picster.itch.io/bottomless): I did not encounter the issue I've encountered on my project, unfortunately. Maybe the color change was so subtle that I didn't recognize it, though I've tried reproducing the issue several times.

The second game was [Blizzard Blowout 64](https://goodgis.itch.io/blizzard-blowout-64): This game had the exact same issue I've had with my project. The colors got brighter all of a sudden that it was hard to miss.

I haven't tested with any Direct3D applications, though I don't remember having such issue with any of them. I may turn off HDR in the mean time, though colors look much better when it's on.
Hope this helps.
